### PR TITLE
Reorganize list widget layout and fix issues

### PR DESCRIPTION
**Pull request by goanpeca**
Part of https://github.com/napari/napari-plugin-manager/issues/14

---

Comments from original issue that were fixed!

1. Align plugin name and summary in the installed plugins section.
2. Align installation info,  Make a standard size?  ( I know this button adjusts by the sizes on the drop down menus held within.  Eliding may help here (see item below).
image

3. Ensure author names are aligned. (maybe misaligned only with many authors and in the presence of an update button?)
image

4. Ensure very long descriptions and author lists are not cut off. 
image

5. Elide version numbers when long. 
When users have a plugin dev version installed the `Installation info` button & section is very long. We probably want to elide this
image

6. refactor code, one example is cleaning up the code that toggles which widgets are visible or not:
https://github.com/napari/napari/pull/5198#discussion_r1028817033

7. Set height of each item.
```
        if item.widget.install_info_button.isExpanded():
            item.widget.setFixedHeight(int(height * SCALE))
        else:
            item.widget.setFixedHeight(int(height / SCALE))
        item.setSizeHint(item.widget.size())
```

---

## Current State with this PR

Screenshot 2024-07-10 at 2 24 58 AM

Screenshot 2024-07-10 at 2 21 44 AM

### CODE REVIEW

**Comment by jaimergp** [comment]
@goanpeca What's the status here? Is this PR still active or should it be closed?

**Comment by goanpeca** [comment]
>@goanpeca What's the status here? Is this PR still active or should it be closed?

Updated the PR, should be ready for review :)

File: napari_plugin_manager/qt_plugin_dialog.py [hunk 1]
```diff
@@ -337,13 +341,79 @@ def setup_ui(self, enabled=True):
...
+        info_layout.addWidget(self.version, 1, 1)
+        self.info_widget.setLayout(info_layout)
+
+        # Error indicator
+        self.error_indicator = QPushButton()
+        self.error_indicator.setObjectName("warning_icon")
+        self.error_indicator.setCursor(Qt.CursorShape.PointingHandCursor)
+        self.error_indicator.hide()
+
+        # --- Layout
```
Comment (jaimergp) [inline napari_plugin_manager/qt_plugin_dialog.py:372]: You can make this a VS Code (and alike) `region` and you get section folding for free 🥳

File: napari_plugin_manager/qt_plugin_dialog.py [hunk 2]
```diff
@@ -235,67 +249,54 @@ def setup_ui(self, enabled=True):
...
+        sizePolicy.setRetainSizeWhenHidden(True)
+        sizePolicy = QSizePolicy(QSizePolicy.Preferred, QSizePolicy.Preferred)
+        self.update_btn.setSizePolicy(sizePolicy)
 
-        self.row2.addWidget(
-            self.update_btn, 0, 6, 1, 1, alignment=Qt.AlignmentFlag.AlignTop
-        )
+        # Action Button
+        self.action_button = QPushButton(self)
+        self.action_button.setFixedWidth(70)
```
Comment (goanpeca) [inline napari_plugin_manager/qt_plugin_dialog.py:289]: Fixes 1, 2, 3

File: napari_plugin_manager/qt_plugin_dialog.py [hunk 3]
```diff
@@ -337,13 +341,79 @@ def setup_ui(self, enabled=True):
...
+        self.info_widget.setLayout(info_layout)
+
+        # Error indicator
+        self.error_indicator = QPushButton()
+        self.error_indicator.setObjectName("warning_icon")
+        self.error_indicator.setCursor(Qt.CursorShape.PointingHandCursor)
+        self.error_indicator.hide()
+
+        # --- Layout
+        # ------------------------------------------------
```
Comment (goanpeca) [inline napari_plugin_manager/qt_plugin_dialog.py:373]: This new layout organization fixes:

1. Align plugin name and summary in the installed plugins section.

File: napari_plugin_manager/qt_plugin_dialog.py [hunk 4]
```diff
@@ -337,13 +341,79 @@ def setup_ui(self, enabled=True):
...
-            7,
-            1,
-            1,
-            alignment=Qt.AlignmentFlag.AlignTop,
+
+        # Information widget for installed packages
+        self.info_widget = QWidget(self)
+        self.info_widget.setLayoutDirection(Qt.LeftToRight)
+        self.info_widget.setObjectName("info_widget")
+        self.info_widget.setFixedWidth(180)
```
Comment (goanpeca) [inline napari_plugin_manager/qt_plugin_dialog.py:349]: Fixes

2. Align installation info, Make a standard size? ( I know this button adjusts by the sizes on the drop down menus held within. Eliding may help here (see item below).

File: napari_plugin_manager/qt_plugin_dialog.py [hunk 5]
```diff
@@ -305,12 +306,11 @@ def setup_ui(self, enabled=True):
...
+        self.install_info_button.setLayoutDirection(
+            Qt.RightToLeft
+        )  # Make icon appear on the right
         self.install_info_button.setObjectName("install_info_button")
-
-        # To make the icon appear on the right
-        self.install_info_button.setLayoutDirection(Qt.RightToLeft)
-
-        # Remove any extra margins
+        self.install_info_button.setFixedWidth(180)
```
Comment (goanpeca) [inline napari_plugin_manager/qt_plugin_dialog.py:313]: Fixes

2. Align installation info, Make a standard size? ( I know this button adjusts by the sizes on the drop down menus held within. Eliding may help here (see item below).

File: napari_plugin_manager/qt_plugin_dialog.py [hunk 6]
```diff
@@ -337,13 +341,79 @@ def setup_ui(self, enabled=True):
...
+        layout_top = QHBoxLayout()
+        layout_bottom = QHBoxLayout()
+        layout_bottom.setSpacing(4)
+
+        layout_left.addWidget(
+            self.enabled_checkbox, alignment=Qt.AlignmentFlag.AlignTop
+        )
+
+        layout_right.addLayout(layout_top, 1)
+        layout_right.addLayout(layout_bottom, 100)
```
Comment (goanpeca) [inline napari_plugin_manager/qt_plugin_dialog.py:387]: Fixes

3. Ensure author names are aligned. (maybe misaligned only with many authors and in the presence of an update button?)

File: napari_plugin_manager/qt_plugin_dialog.py [hunk 7]
```diff
@@ -235,67 +249,54 @@ def setup_ui(self, enabled=True):
...
-        self.row2.setSpacing(4)
+        # Summary
         self.summary = QElidingLabel(parent=self)
         self.summary.setObjectName('summary_text')
         self.summary.setWordWrap(True)
 
-        sizePolicy = QSizePolicy(QSizePolicy.Preferred, QSizePolicy.Preferred)
+        font_summary = QFont()
+        font_summary.setPointSize(10)
+        self.summary.setFont(font_summary)
```
Comment (goanpeca) [inline napari_plugin_manager/qt_plugin_dialog.py:265]: Fixes

4. Ensure very long descriptions and author lists are not cut off.

File: napari_plugin_manager/qt_plugin_dialog.py [hunk 8]
```diff
@@ -337,13 +341,79 @@ def setup_ui(self, enabled=True):
...
+        # Information widget for installed packages
+        self.info_widget = QWidget(self)
+        self.info_widget.setLayoutDirection(Qt.LeftToRight)
+        self.info_widget.setObjectName("info_widget")
+        self.info_widget.setFixedWidth(180)
+
+        self.source_text = QLabel('Source:')
+        self.source = QLabel(PYPI)
+        self.version_text = QLabel('Version:')
+        self.version = QElidingLabel()
```
Comment (goanpeca) [inline napari_plugin_manager/qt_plugin_dialog.py:354]: Fixes 

5. Elide version numbers when long.
When users have a plugin dev version installed the Installation info button & section is very long. We probably want to elide this

File: napari_plugin_manager/qt_plugin_dialog.py [hunk 9]
```diff
@@ -133,23 +136,29 @@ def __init__(
...
         icon = QColoredSVGIcon.from_resources('logo_silhouette')
-        npe2_icon.setPixmap(
-            icon.colored(color='#33F0FF', opacity=opacity).pixmap(20, 20)
+        self.set_status(
+            icon.colored(color='#33F0FF', opacity=opacity).pixmap(20, 20), text
         )
-        self.row1.insertWidget(2, QLabel(lbl))
-        self.row1.insertWidget(2, npe2_icon)
+
+    def set_status(self, icon=None, text=''):
```
Comment (goanpeca) [inline napari_plugin_manager/qt_plugin_dialog.py:152]: Fixes 

6. refactor code, one example is cleaning up the code that toggles which widgets are visible or not:
https://github.com/napari/napari/pull/5198#discussion_r1028817033

File: napari_plugin_manager/qt_plugin_dialog.py [hunk 10]
```diff
@@ -455,6 +488,7 @@ def __init__(self, parent: QWidget, installer: InstallerQueue) -> None:
         self.setSortingEnabled(True)
         self._remove_list = []
         self._data = []
+        self._initial_height = None
```
Comment (goanpeca) [inline napari_plugin_manager/qt_plugin_dialog.py:491]: Fixes

7. Set height of each item.

File: napari_plugin_manager/qt_plugin_dialog.py [hunk 11]
```diff
@@ -549,18 +583,20 @@ def addItem(
         )
 
         item.setSizeHint(item.widget.size())
+        if self._initial_height is None:
+            self._initial_height = item.widget.size().height()
```
Comment (goanpeca) [inline napari_plugin_manager/qt_plugin_dialog.py:587]: Fixes

7. Set height of each item.

File: napari_plugin_manager/qt_plugin_dialog.py [hunk 12]
```diff
@@ -549,18 +583,20 @@ def addItem(
...
         widg.install_info_button.toggled.connect(
             lambda: self._resize_pluginlistitem(item)
         )
 
     def _resize_pluginlistitem(self, item):
         """Resize the plugin list item, especially after toggling QCollapsible."""
-        height = item.widget.height()
         if item.widget.install_info_button.isExpanded():
-            item.widget.setFixedHeight(int(height * SCALE))
+            item.widget.setFixedHeight(self._initial_height + 35)
```
Comment (goanpeca) [inline napari_plugin_manager/qt_plugin_dialog.py:597]: Fixes

7. Set height of each item.

**Comment by goanpeca** [comment]
@jaimergp added the comments :)

**Comment by jaimergp** [comment]
Thanks @goanpeca. @dalthviz, can you confirm the proposed changes address the items presented in https://github.com/napari/napari-plugin-manager/issues/14? Thanks!

**Comment by dalthviz** [comment]
Gave this a check on Windows and seems like the listed items from #14 that are being mentioned here get a fix here indeed :+1: However, not totally sure if the behavior of the layout after some interactions with the `Installation Info` widget and the way the version label is getting elided are expected. Just in case, a preview of how those things were behaving for me:

plugin_dialog_layout

**Comment by jaimergp** [comment]
Yea, looks like the extra button in the top panel (Update X.Y.Z) makes things behave a bit differently 🤔

**Comment by goanpeca** [comment]
Eliding indeed is no working as expected, definitely a bug on my side :)

Thanks for the review @dalthviz indeed weird behavior 😆 on resize, taking a look at it now.

**Comment by psobolewskiPhD** [comment]
Only thing I noticed was a flicker when you filter while it's still fetching:
image
Screenshot doesn't show it well, but type zarr into filter, the description of the plugin flickers.

**Comment by goanpeca** [comment]
@dalthviz added some dummy version to test the elide I think it is working, it just needs to be longer to be elided. Also added a fix for the resizing behavior bug you found. Could you check again :) ?

napari

**Comment by dalthviz** [comment]
Gave this another check and seems like the resize behavior I mentioned above is not reproducible anymore :+1: I still see the version string being cut in two lines from my side but seems like that is expected due to the version string length, if I'm understanding correctly, right?:

plugin_resize_elide_behavior

Also, related with the flickering behavior, I tried filtering things multiple times while the available plugins list seemed like still being populated. For comparison, couple of GIF showing the behaviors (changes here vs latest release - 0.1.0a2 vs latest main - 5027417):

| Version | Preview |
|-|-|
|This PR|plugins_filter|
|`0.1.0a2`|plugins_filter_010a2|
|main (5027417)|plugins_filter_main|

I see with the changes here and with latest main a little freeze when filtering things, maybe is worthy to create an issue to track that? Related more specifically with the flickering, I think I'm able to see that behavior in all the versions that I checked. Maybe that is related with changes caused due to other plugins being loaded which change the width used for other fields like author and then cause the plugin description space to be shorter? Maybe creating an issue for that could be worthy too? 🤔

**Comment by goanpeca** [comment]
Hi @dalthviz thanks for the check again. Now I understand that the version is split in 2 lines... which is odd but, I know how to fix it :) 

Will push a fix. Thanks again for the thorough reviews¡

**Comment by goanpeca** [comment]
@dalthviz so, the issue is that the eliding logic which is based for text, will split on given characters like a dot. That is why you see the version split in 2 lines. So one "solution" would be to replace a normal dot by a different type of symbol. I tried a few and this middle dot one seems to work but I am not sure if it would make the version look too weird. It does look odd to me, but I could not find another dot symbol that did not end up splitting things as well.
 
Thoughts @jaimergp ?


Screenshot 2024-07-17 at 5 19 31 PM

Screenshot 2024-07-17 at 5 21 20 PM

**Comment by jaimergp** [comment]
> So one "solution" would be to replace a normal dot by a different type of symbol. I tried a few and found one that could work :)

lmao hahah love it

**Pull request merged**